qucs-core/src/components/taperedline.cpp

```diff
@@ -0,0 +1,268 @@
+  nr_complex_t gamma = nr_complex_t (alpha, beta); //Complex propagation constant
+  matrix ABCD_ = eye(2);//Overall ABCD matrix
+  matrix ABCDaux = eye(2);//Auxiliar matrix for performing the iterative product
+  nr_double_t Zi;
+  nr_complex_t a, b, c, d;
+  // ABCD coefficients
+  a = cosh(gamma*lstep);
+  b = sinh(gamma*lstep); // need to be multiplied by Zi
+  c = sinh(gamma*lstep); // need to be divided by Zi
+  d = cosh(gamma*lstep); 
+
+  nr_double_t l = lstep/2.0; // compute impedance in the middle of the section
+  for (int idx = 0 ; idx < Nsteps; idx++)
+  {
+    // The line is discretized in finite elements. The size of these elements can be considered a differential
+    // length since the impedance change across the actual section is small. 
+    // Taking into account the cascading property of the ABCD matrix, the overall
+    // ABCD matrix can be calculated as the product of the individual ABCD matrices.   
+    Zi = Zprofile[idx];
+    ABCDaux.set(0,0,a);
+    ABCDaux.set(0,1,Zi*b);
+    ABCDaux.set(1,0,c/Zi);
+    ABCDaux.set(1,1,d);
+    //Iterative product
+    ABCD_=ABCD_*ABCDaux;
+    l += lstep; // move to the next section
+  }
+  // Overall ABCD coefficients
+  ABCD = ABCD_;
+}
+
+//------------------------------------------------------------------
+// Exponential impedance profile
+nr_double_t taperedline::calcExponential(nr_double_t l, nr_double_t L, nr_double_t Z1, nr_double_t Z2)
+{
+   nr_double_t a = (1/L)*std::log(Z2/Z1);
+   return Z1*std::exp(a*l);
+}
+
+//------------------------------------------------------------------
+// Linear impedance profile
+nr_double_t taperedline::calcLinear(nr_double_t l, nr_double_t L, nr_double_t Z1, nr_double_t Z2)
+{
+   return Z1+l*(Z2-Z1)/L;
+}
+
+//------------------------------------------------------------------
+// Triangular impedance profile
+nr_double_t taperedline::calcTriangular(nr_double_t l, nr_double_t L, nr_double_t Z1, nr_double_t Z2)
+{
+   if (l < L/2)
+   {
+     return Z1*std::exp((2*l*l/(L*L))*std::log(Z2/Z1));
+   }
+
+   return Z1*std::exp(((4*l/L) - (2*l*l/(L*L)) -1)*std::log(Z2/Z1));
+}
+
+//------------------------------------------------------------------
+// Klopfenstein impedance profile
+nr_double_t taperedline::calcKlopfenstein(nr_double_t l, nr_double_t L, nr_double_t Z1, nr_double_t Z2, nr_double_t gamma_max)
+{
+   nr_double_t gamma0 = 0.5*std::log(Z2/Z1);
+   nr_double_t A = std::acosh(gamma0/gamma_max);
+   return (std::exp(0.5*std::log(Z1*Z2) + (gamma0/std::cosh(A))*A*A*phi(2.*(l/L)-1., A)));
+}
+
+void taperedline::calcImpedanceProfile()
+{
+nr_double_t L = getPropertyDouble ("L");//Length
+  nr_double_t Z1 = getPropertyDouble ("Z1");//Port 1 impedance
+  nr_double_t Z2 = getPropertyDouble ("Z2");//Port 2 impedance
+
+  if (Z1 > Z2)//Handling Z1 > Z2 case
+  {
+     logprint (LOG_ERROR, "WARNING: The impedance at port 1 is bigger than the impedance at port 2 ((Z1 = %g Ohm ) > (Z2 = %g Ohm))\n", Z1, Z2);
```

Comment:
uhm, OK for the warning but if Z1 > Z2 the results are wrong. You are exchanging the ports impedances but not the actual ports connections. Either simply throw an error for this case or handle it by also swapping the two ports (you could use something like S.exchangeRows(0,1); S.exchangeCols(0,1))